### PR TITLE
First branch

### PR DESCRIPTION
**Pull request by danikkrapivnitskiy**

### CODE REVIEW

File: JSTutorial/hw_8/task_4/Task.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+class Task extends ITSpecialist {
```
Comment (VladimirLQA) [inline JSTutorial/hw_8/task_4/Task.js]: Why `Task` extends `ITSpecialist` ?

File: JSTutorial/hw_8/task_4/Task.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    #featureName
+    #userStoryNumber
+    #estimations
+
+    constructor(featureName, userStoryNumber, estimations) {
+        super();
+        this.featureName = featureName
+        this.userStoryNumber = userStoryNumber
```
Comment (VladimirLQA) [inline JSTutorial/hw_8/task_4/Task.js]:
```suggestion
        this.#featureName = featureName
        this.#userStoryNumber = userStoryNumber
```

File: JSTutorial/hw_8/task_4/Team.js [hunk 3]
```diff
@@ -0,0 +1,103 @@
+class Team {
+    #name
+    #sprint_duration
+    #release_date
+    #daily_standup_time
+    #teammates = []
+    #tasks = []
+
+    constructor(name, sprint_duration, release_date, daily_standup_time, teammates, tasks) {
```
Comment (VladimirLQA) [inline JSTutorial/hw_8/task_4/Team.js:9]: Private fields need to be initialized with the `#` prefix in the constructor

File: JSTutorial/hw_8/task_4/Team.js [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    addTeammate(newTeammate) {
+        if (!(newTeammate instanceof ITSpecialist)) {
+            throw new Error("Invalid teammate");
+        }
+        this.#teammates.push(newTeammate);
+    }
+
+    removeTeammate({name}) {
```
Comment (VladimirLQA) [inline JSTutorial/hw_8/task_4/Team.js]:
```suggestion
    removeTeammate(name) {
```

File: JSTutorial/hw_7/task_2.js [hunk 5]
```diff
@@ -216,16 +219,16 @@ addEnterprise("Предприятие 4")
...
     return enterpriseObj;
 }
 function addDepartment(id, name) {
-    const enterprise = getEnterpriseByIdOrName(id);
-    const newObj = {
+    const enterprise = getEnterpriseByIdOrName(id, enterprises);
+    const nejObj = {
         name,
         employees_count: [],
         id : enterprise.id + 12,
```
Comment (anatoly-karpovich) [inline JSTutorial/hw_7/task_2.js]: Where is function that creates new id?)

File: JSTutorial/hw_8/task_4/ITSpecialist.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
+class ITSpecialist {
+    #name
+    #grade
+    #experience_in_years
+    #age
+    #country
+    #_salary
```
Comment (anatoly-karpovich) [inline JSTutorial/hw_8/task_4/ITSpecialist.js]:
```suggestion
    #salary
```
You are already using private field, no need to add underscore

File: JSTutorial/hw_8/task_4/Manager.js [hunk 7]
```diff
@@ -0,0 +1,10 @@
+class Manager extends ITSpecialist{
+    #isScrumMaster
+    constructor() {
```
Comment (anatoly-karpovich) [inline JSTutorial/hw_8/task_4/Manager.js]: Constructor of class should have at least the same arguments as a parent class + it's own. And you should use arguments, related to parent class, in super(parent class arguments)

File: JSTutorial/hw_8/task_4/QA.js [hunk 8]
```diff
@@ -0,0 +1,10 @@
+class QA extends ITSpecialist{
+    #isAqa;
+    constructor() {
```
Comment (anatoly-karpovich) [inline JSTutorial/hw_8/task_4/QA.js]: Same as for mamager

File: JSTutorial/hw_8/task_4/Task.js (hunk 1 above)
Comment (anatoly-karpovich) [inline JSTutorial/hw_8/task_4/Task.js]: How a task can extend ITSpecialist? Does task have name, country or other properties?)

File: JSTutorial/hw_8/task_4/Manager.js [hunk 9]
```diff
@@ -0,0 +1,10 @@
...
+        super();
+        this.#isScrumMaster = true;
```
Comment (anatoly-karpovich) [inline JSTutorial/hw_8/task_4/Manager.js]: This should be passes to constructor as well, not just hardcoded aa true

File: JSTutorial/hw_8/task_4/Team.js [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    showAllTeammates() {
+        return this.#teammates;
+    }
+
+    showTeammatesBySpecialization(specialization) {
+        let predicate;
+        switch (specialization) {
+            case "developer" : predicate = team => team.isWritingUnitTests(); break;
```
Comment (anatoly-karpovich) [inline JSTutorial/hw_8/task_4/Team.js]: isWritingUnitTests() can return false. Maybe it's better to check with instanceof?)

File: JSTutorial/hw_8/task_4/Team.js [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+            case "aqa" : predicate = team => team.isAqa(); break;
+            case "manager" : predicate = team => team.isScrumMaster(); break;
+        }
+        if (predicate !== undefined) {
+            return this.#teammates.filter(team => team.name() === name);
```
Comment (anatoly-karpovich) [inline JSTutorial/hw_8/task_4/Team.js]: I don't see any initialisation of "name" variable in this method. What is it?

**Review by anatoly-karpovich — APPROVED** [review]

File: JSTutorial/hw_8/task_4/Team.js [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+            specialist[key] = value;
+        } else throw new Error("Specialist not found")
+    }
+
+    showAllTeammates() {
+        return this.#teammates;
+    }
+
+    showTeammatesBySpecialization(specialization) {
+        let predicate;
```
Comment (VladimirLQA) [inline JSTutorial/hw_8/task_4/Team.js]: We can provide `specialization` field and use it for search
```js
          return this._teammates.filter((teammate) => teammate.specialization === specialization);

```

File: JSTutorial/hw_8/task_4/Team.js [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+            throw new Error("Invalid teammate");
+        }
+        this.#teammates.push(newTeammate);
+    }
+
+    removeTeammate({name}) {
+        if (typeof name !== "string") {
+            throw new Error("Invalid name type");
+        }
+        this.#teammates.splice(this.#teammates.findIndex(specialist => specialist.name() === name), 1);
```
Comment (VladimirLQA) [inline JSTutorial/hw_8/task_4/Team.js:70]: Do not use brackets for getters / setters
https://javascript.info/property-accessors#getters-and-setters

File: JSTutorial/hw_8/task_4/Team.js [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+            case "manager" : predicate = team => team.isScrumMaster(); break;
+        }
+        if (predicate !== undefined) {
+            return this.#teammates.filter(team => team.name() === name);
+        } else throw new Error("Specialization not found")
+    }
+
+    showAllTasks() {
+        return this.#teammates.map(teammate => teammate.featureName());
+    }
```
Comment (VladimirLQA) [inline JSTutorial/hw_8/task_4/Team.js]:
```suggestion
    showAllTasks() {
        return this.#tasks;
    }
```

**Review by VladimirLQA — APPROVED** [review]

File: JSTutorial/hw_8/task_2.js [hunk 15]
```diff
@@ -8,14 +8,21 @@
 
 //1
 function counter(n) {
-    let start = 1;
-    let counter = n;
+    let arr = new Set;
```
Comment (VladimirLQA) [inline JSTutorial/hw_8/task_2.js:11]:
```suggestion
    const arr = new Set();
```

File: JSTutorial/hw_8/task_4/Team.js (hunk 12 above)
Comment (danikkrapivnitskiy) [reply JSTutorial/hw_8/task_4/Team.js]: why? Specialization is not a function, it's just a key

**Pull request merged**